service_config.md

```diff
@@ -0,0 +1,194 @@
+   by the Pod                                       |                                                                                    |
+3. Kolla sync src / dest and applies the expected   | [keystone_authtoken]                                                               |
+   ownership/permissions                            | password = {{.Password}}                                                           |
+                                                    |                                                                                    |
+                     +----------------------------- |                                                 (1)                                |
+                     |                              | [service_user]                                                                     |
+                     |                              | password = {{.Password}}                                                           |
+                     |                              +------------------------------------------------------------------------------------+
+                     |
++----------------------------------------+      +----------------------------------------------------------------------------------------+
+| apiVersion: v1                   (2)   |      |  “config_files”: [        (3)                                                          |
+| metadata:                              |      |                  ...                                                                   |
+|   name: 01-<service>-deployment-secret |      |      {                                                                                 |
+|   namespace: openstack                 | ===> |        "source": "/var/lib/config-data/deployment/01-deployment-<service>-secret.conf",|
+| stringData:                            |      |        "dest": "/etc/<service>/<service>.conf.d/01-deployment-<service>-secret.conf",  |
+|   01-deployment-<service>.conf: |      |      |        "owner": "<service>",                                                           |
+|        <data>                          |      |        "perm": "0600"                                                                  |
++----------------------------------------+      |      },                                                                                |
+                                                +----------------------------------------------------------------------------------------+
+```
+
+As the diagram above depicts, a given operator is supposed to implement the
+logic to build and reconcile the `Secret` when the template has been processed.
+The `Secret`, mounted to the resulting service Pod, in the very last step of
+this process is synced by kolla to the destination directory: instead of having
```

Comment:
directly mounted in /etc/cinder/cinder.conf.d hides all the original contents of that dir, exposing only files from the secret. That could become a problem for minor updates, for example, or gradual passwords rotation (where we might want to have both old config with the old password  alongside the updated config with the changed credentials). So we'' need to use subPaths

Reply:
What original contents of the directory?  The directory is empty and the defaults come from the operator.
How would you envision gradual passwords rotation working?  Because I think I'm missing something.

Reply:
@SeanMooney ^^ fyi. @Akrog please see [OSPRH-92](https://issues.redhat.com/browse/OSPRH-92) that explains the "gradual passwords rotation" w/o openstack services downtime. Does that also apply for Cinder backends access creds? If so, the backend config directory *might* need to maintain the "old" version of credentials, and the new (rotated) ones in another set of configs for some transitory time?

wrt "hidden" contents in cinder.conf.d: "As we mount a Kubernetes volume at a mount-point, and if we already have some content in it, then it would be hidden by our mount. But in some cases, we would not want to hide the existing content at the mount-point but want to add additional files/directories in parallel"

Reply:
> @SeanMooney ^^ fyi. @Akrog please see [OSPRH-92](https://issues.redhat.com/browse/OSPRH-92) that explains the "gradual passwords rotation" w/o openstack services downtime. Does that also apply for Cinder backends access creds? If so, the backend config directory _might_ need to maintain the "old" version of credentials, and the new (rotated) ones in another set of configs for some transitory time?

Sorry @bogdando, I still don't see how you envision the credentials rotation working...
Cinder services cannot use 2 credentials at the same time, and I doubt many OpenStack services can.
Once we decide to change credentials the services need to be restarted (which we can probably do progressively) with the new configuration.
So during the rotation we'll have services using the old credentials and some using the new credentials, and if both are valid then everything is fine.  At some point all services will be using the new credentials and the old credentials can be "deleted".

Am I mistaken in thinking that's how the rotation would work?


> wrt "hidden" contents in cinder.conf.d: "As we mount a Kubernetes volume at a mount-point, and if we already have some content in it, then it would be hidden by our mount. But in some cases, we would not want to hide the existing content at the mount-point but want to add additional files/directories in parallel". But if the directory is empty and all comes from the operator, that shouldn't be a problem

For cinder we don't currently use that together, and even if we were we don't want the RPM defaults, we want the defaults that are defined in our operator.  I thought we had all agreed that operators would have their own default configuration files stored in their repositories and would not use any other defaults, but I could be wrong here.

Reply:
yes, that makes sense